containermetrics/cpu_spike_handler.go

```diff
@@ -0,0 +1,73 @@
+package containermetrics
+
+import (
+	"time"
+
+	loggingclient "code.cloudfoundry.org/diego-logging-client"
+	"code.cloudfoundry.org/executor"
+	"code.cloudfoundry.org/lager"
+)
+
+type spikeInfo struct {
+	start *time.Time
+	end   *time.Time
+}
+
+type CPUSpikeHandler struct {
+	spikeInfos   map[string]*spikeInfo
+	metronClient loggingclient.IngressClient
+}
+
+func NewCPUSpikeHandler(metronClient loggingclient.IngressClient) *CPUSpikeHandler {
+	return &CPUSpikeHandler{
+		spikeInfos:   make(map[string]*spikeInfo),
+		metronClient: metronClient,
+	}
+}
+
+func (handler *CPUSpikeHandler) Handle(logger lager.Logger, container executor.Container, metric executor.Metrics, timeStamp time.Time) error {
+	guid := container.Guid
+	previousSpikeInfo := handler.spikeInfos[guid]
+	currentSpikeInfo := &spikeInfo{}
+
+	if previousSpikeInfo != nil {
+		currentSpikeInfo.start = previousSpikeInfo.start
+		currentSpikeInfo.end = previousSpikeInfo.end
+	}
+
+	if spikeStarted(metric, previousSpikeInfo) {
+		currentSpikeInfo.start = &timeStamp
+		handler.spikeInfos[guid] = currentSpikeInfo
+		return nil
+	}
+
+	if spikeEnded(metric, previousSpikeInfo) {
+		currentSpikeInfo.end = &timeStamp
+
+		err := handler.metronClient.SendSpikeMetrics(loggingclient.SpikeMetric{
+			Start: *currentSpikeInfo.start,
+			End:   *currentSpikeInfo.end,
+			Tags:  metric.MetricsConfig.Tags,
+		})
+		if err != nil {
+			return err
+		}
+
+		handler.spikeInfos[guid] = nil
```

Comment:
Maybe better to use delete(handler.spikeInfos, guid) to add some prevention for map not growing in size. 

If container is going away though this key is not going to get removed. So maybe there should be a clean up of all the keys that are not currently spiking.

Reply:
Yeah, the existing stats reporter uses an `atomic.Value` as the metrics cache and each run of the reporter recreates the map that is stored. We could use a similar pattern for this one instead of adding to a forever-growing map.

This spike/changes on the other stats reporter have the same issue. The `repMetricsMap` (and therefore `metricsCache`) is created in the constructor and never reset.